File: src/com/anathema_roguelike/environment/LocationProperty.java
/*******************************************************************************
 * Copyright (C) 2017 Carl Minden
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package com.anathema_roguelike.environment;

import com.anathema_roguelike.main.Game;
import com.anathema_roguelike.main.display.DungeonMap.DungeonLayer;
import com.anathema_roguelike.main.display.VisualRepresentation;
import com.anathema_roguelike.stats.effects.Effect;
import com.anathema_roguelike.stats.effects.HasEffect;
import com.anathema_roguelike.stats.locationstats.LocationStat;

public abstract class LocationProperty implements HasEffect<Effect<Location, LocationStat>>, HasLocation {
	
	private VisualRepresentation visualRepresentation;
	private VisualRepresentation fogOfWarRepresentation;
	private Location location;
	
	private boolean passable;
	private double opacity;
	private double damping;
	private double brightness = 0;
	
	private DungeonLayer layer;
	private DungeonLayer fogOfWarLayer;
	
	public LocationProperty(VisualRepresentation representation, VisualRepresentation fogOfWarRepresentation, boolean foreground, boolean passable, double opacity, double damping) {
		this.passable = passable;
		this.visualRepresentation = representation;
		this.fogOfWarRepresentation = fogOfWarRepresentation;
		this.opacity = opacity;
		this.damping = damping;
		
		if(foreground) {
			layer = DungeonLayer.FOREGROUND;
			fogOfWarLayer = DungeonLayer.FOG_OF_WAR_FOREGROUND;
		} else {
			layer = DungeonLayer.BACKGROUND;
			fogOfWarLayer = DungeonLayer.FOG_OF_WAR_BACKGROUND;
		}
		
	}
	
	@Override
	public Location getLocation() {
		return location;
	}
	
	public void setLocation(Location location) {
		this.location = location;
	}
	
	public boolean isPassable() {
		return passable;
	}
	
	public double getOpacity() {
		return opacity;
	}
	
	public double getDamping() {
		return damping;
	}
	
	public double getBrightness() {
		return brightness;
	}
	
	protected void setBrightness(double brightness) {
		this.brightness = brightness;
	}
	
	protected void renderToFogOfWar() {
		Game.getInstance().getMap().renderVisualRepresentation(fogOfWarLayer, location.getX(), getY(), getFogOfWarRepresentation());
	}
	
	protected void render() {
		Game.getInstance().getMap().renderVisualRepresentation(layer, getX(), getY(), getRepresentation());
		
		renderToFogOfWar();
	}

	public VisualRepresentation getRepresentation() {
		return visualRepresentation;
	}
	
	public VisualRepresentation getFogOfWarRepresentation() {
		if(fogOfWarRepresentation != null) {
			return fogOfWarRepresentation;
		} else {
			return getRepresentation();
		}
	}
}

File: src/com/anathema_roguelike/environment/LocationProperty.scala


package com.anathema_roguelike
package environment

import com.anathema_roguelike.main.Game
import com.anathema_roguelike.main.display.DungeonMap.DungeonLayer
import com.anathema_roguelike.main.display.VisualRepresentation
import com.anathema_roguelike.stats.Stat.LocationStat
import com.anathema_roguelike.stats.effects.{Effect, HasEffect}
abstract class LocationProperty(
    location: Location,
    visualRepresentation: VisualRepresentation,
    fogOfWarRepresentation: VisualRepresentation,
    foreground: Boolean,
    passable: Boolean,
    opacity: Double,
    damping: Double) extends HasEffect[Effect[Location, LocationStat]] with HasLocation {

  private var brightness = 0.0

  private val layer = if(foreground) DungeonLayer.FOREGROUND else DungeonLayer.FOG_OF_WAR_FOREGROUND

  private val fogOfWarLayer = if(foreground) DungeonLayer.FOG_OF_WAR_FOREGROUND else DungeonLayer.FOG_OF_WAR_BACKGROUND

  override def getLocation: Location = location

  def isPassable: Boolean = passable

  def getOpacity: Double = opacity

  def getDamping: Double = damping

  def getBrightness: Double = brightness

  protected def setBrightness(brightness: Double): Unit = {
    this.brightness = brightness
  }

  protected def renderToFogOfWar(): Unit = {
    Game.getInstance.getMap.renderVisualRepresentation(fogOfWarLayer, location.getX, getY, getFogOfWarRepresentation)
  }

  def render(): Unit = {
    Game.getInstance.getMap.renderVisualRepresentation(layer, getX, getY, getRepresentation)
    renderToFogOfWar()
  }

  def getRepresentation: VisualRepresentation = visualRepresentation

  def getFogOfWarRepresentation: VisualRepresentation = if (fogOfWarRepresentation != null) fogOfWarRepresentation
  else getRepresentation
}
